### ISSUE TITLE
utf8-chars limited to <65535 (brl.blitz and brl.stringbuilder)

### ISSUE THREAD

**Issue by GWRon** [bmx-ng/brl.mod#149]
As posted in
https://www.syntaxbomb.com/index.php/topic,5439.msg28403/topicseen.html#new

the current implementation of `bbStringFromUTF8String` in "blitz.mod/blitz_string.c" and `bmx_stringbuilder_append_utf8string` in "stringbuilder.mod/glue.c" throw an exception if they receive an utf8 character with a code bigger than 65535 (unsigned short limit).

Affected characters are eg. "unicode emojis".

The current implementation uses a "short buffer" which has this limit and therefor does not accept bigger unicode characters.

Maybe https://github.com/openssl/openssl/blob/master/crypto/asn1/a_utf8.c
`int UTF8_getc(const unsigned char *str, int len, unsigned long *val)` might be worth a look as it uses a long buffer (and some more stuff which might be useful to have?).

**Comment by woollybah** [comment]
BlitzMax characters are "Short" bytes in size. This is by design. Without significantly changing the design, there's not much that can be done about it.

**Comment by GWRon** [comment]
Hmm maybe the code could gain an option to:
Add the html code instead...
or a default replacement character?

So a way to eg. use brl.json to fetch and process stuff without running into the exception or loosing data/text information.

**Comment by woollybah** [comment]
BlitzMax doesn't support characters over a certain range. There's not much you can do about it.

**Comment by GWRon** [comment]
Even not when converting from some external source / stream?

A pity.
You might chime in in the syntaxbomb thread then.

**Comment by woollybah** [comment]
Well, converting from some source to what? A smiley has a value of 128512. That simply doesn't fit into 65535.

**Comment by GWRon** [comment]
Yes it does not fit...but could be converted to some escaped expression \uedef or an hex value. Or a default "??" character if desired.
Better than throwing an exception when using brl.json to read some api response containing chat messages (regarding to the thread posts).

**Comment by GWRon** [comment]
If it was for brl.json only then the functions could accept a callback as param. The callback is used to handle unsupported data...to eg. convert it into html codes or ascii equivalents.

For the generic brl.blitz function I am not sure about the impact of an passed function pointer ...but as it is only executed for chars above the "short limit" it should be ok too (for "normal texts")

**Comment by Hezkore** [comment]
Smileys do seem to fit in strings.
It's only when you do stuff like FromUTF8String it seems to throw an error.
```
SuperStrict
Framework BRL.StandardIO

'Local emojiString:String =  String.FromUTF8String( "💩" ) ' < Doesn't work
Local emojiString:String =  "💩" ' < Works

' Opening the result in Sublime shows the emoji
SaveString( emojiString, "tmp.txt"  )
```

**Comment by noroads2015** [comment]
been looking at this one and I found a possible solution.  UCS-2 uses what are called surrogate pairs.  This is a way to encode values above 65535 into 2 16 bit UCS characters.  If you replace brl.mod/blitz.mod/blitz_string.c  lines 267-268 with this, then any UTF-8 code above 65535 wil be encoded into a surrogate pair.  Of course, if you try and print the result in BlitzMax, you will only get random characters as BlitzMax strings doesn't support code points above 65535, but you will still be able to process the string, and BlitzMax won't crash.
```
Function SurrogateToCodePoint:Int(High:Int, Low:Int)
	If (High & $d800) <> $d800 Then Return -1 'Not a valid high surrogate
	If (Low & $dc00) <> $dc00 Then Return -1 'not a valid low surrogate
	
	Local value:Int = ((High - $d800) Shl 10) + (Low - $dc00) + $10000
	Return value
End Function
```

**Comment by woollybah** [comment]
Yeah, that would be part of a solution, and perhaps would be fine just to handle storing large unicode values within the String.

You also need to take into account the fact that you may need more space to store the text now. Plus, you probably want to be able to get from your new String back to UTF8 - so that the unicode is encoded within the UTF8 string correctly.

So, if you are going to go to that amount of effort, then maybe it's worth more work on top of that to do it properly - for example, should String length return the number of "characters" or the number of "shorts" ? Currently, that answer is the same, but supporting surrogate pairs would give different answers to that question...

**Comment by GWRon** [comment]
Maybe string.size() could return the bytes and string.length() the amount of "entries" (so characters). This way it behaved the same as an array.length call.

**Comment by noroads2015** [comment]
No worries about text storage, as the function takes the length of the UTF-8 string in bytes and doubles it.  Since UTF-8 values above 65535 are encoded in 4 bytes, then 4 shorts are reserved for it, but only 2 are needed.  So extra space is actually created.

Now going from Unicode to UTF-8 would require allocating a bit of additional space.  Currently, it allocates 3 bytes per character, but you would need to allocate 4 bytes per character to be sure that enough space is created.

As for String.Length, it should still report the number of shorts.  Don't want to break legacy behavior.  Might be a good idea to add a String.charLength property or something to give the actual number of characters.

**Comment by noroads2015** [comment]
Here is the modified bbStringFromUTF8String() and bbStringToUTF8String() functions.  Tested them and they seem to be working properly.  Unfortunately, SaveText() and LoadText() use their own ITF-8<->Blitz String functions, so would need to change those next.
bbStringFromUTF8String()
```
BBString *bbStringFromUTF8String( const char *p ){
	int c,n;
	short *d,*q;
	BBString *str;

	if( !p ) return &bbEmptyString;
	
	n=strlen(p);
	d=(short*)malloc( n*2 );
	q=d;
	
	while( c=*p++ & 0xff ){
		if( c<0x80 ){
			*q++=c;
		}else{
			int d=*p++ & 0x3f;
			if( c<0xe0 ){
				*q++=((c&31)<<6) | d;
			}else{
				int e=*p++ & 0x3f;
				if( c<0xf0 ){
					*q++=((c&15)<<12) | (d<<6) | e;
				}else{
					int f=*p++ & 0x3f;
					int v=((c&7)<<18) | (d<<12) | (e<<6) | f;
					if( v & 0xffff0000 ) //bbExThrowCString( "Unicode character out of UCS-2 range" );
					{
						v -= 0x10000;
						d = ((v >> 10)& 0x7ffff) + 0xd800;
						e = (v & 0x3ff) + 0xdc00;
						*q++=d;
						*q++=e;
					}else{
						*q++=v;
					}
				}
			}
		}
	}
	str=bbStringFromShorts( d,q-d );
	free( d );
	return str;
}
```
bbStringToUTF8String()
```
char *bbStringToUTF8String( BBString *str ){
	int i,len=str->length;
	char *buf=(char*)bbMemAlloc( len*4+1 );
	char *q=buf;
	unsigned short *p=str->buf;
	for( i=0;i<len;++i ){
		unsigned int c=*p++;
		if(0xd800 <= c && c <= 0xdbff) {
			/* surrogate pair */
			unsigned int c2 = *p;
			if(0xdc00 <= c2 && c2 <= 0xdfff) {
				/* valid second surrogate */
				c = ((c - 0xd800) << 10) + (c2 - 0xdc00) + 0x10000;
				p++;
			}
		}
			
		if( c<0x80 ){
			*q++=c;
		}else if( c<0x800 ){
			*q++=0xc0|(c>>6);
			*q++=0x80|(c&0x3f);
		}else if( c<0x10000){
			*q++=0xe0|(c>>12);
			*q++=0x80|((c>>6)&0x3f);
			*q++=0x80|(c&0x3f);
		}else if( c< 0x10ffff){
			*q++=0xf0|(c>>18);
			*q++=0x80|((c>>12)&0x3f);
			*q++=0x80|((c>>6)&0x3f);
			*q++=0x80|(c&0x3f);
		}else{
			bbExThrowCString( "Unicode character out of UTF-8 range" );
		}
	}
	*q=0;
	return buf;
}
```

**Comment by noroads2015** [comment]
Looked at the latest commit for blitz_string.c.  There are a couple of potential issues.  First, I made a small mistake in my code that you copied.  In bbStringFromUTF8String, the line
d = ((v >> 10)& 0x7ffff) + 0xd800;
should be 
d = ((v >> 10)& 0x7ff) + 0xd800;
You are masking 11 bits, not 19.  It really shouldn't be a problem, as the current Unicode specifications only goes up to 21 bits.

The second issue is a bit more problematic.  In my version of bbStringToUTF8String function, if it comes across one of the high surrogate values ($D800-DBFF), it then checks to see if the next value is a low surrogate ($DC00-$DFFF).  If not, then it encodes the two values separately into the UTF8 stream without decoding the codepoint first. Blitzmax strings can potentially hold any kind of value.  Maybe level data, or sensor readings, etc... Not just text characters.  So it would be important to check for valid surrogate pairs.
In your version, it just assumes the second value is a valid surrogate, and tries to decode it.  If it is not valid, then the value will just get mangled, and when you try and load the value back into a BMax string, it will be incorrect.

Your version has one advantage over mine, if the high surrogate happens to be the last character of the string, it will create an error.  My version might run normally, add an extra character, or MAV depending on what is following the string.  Easy fix though, just change line 776 in my version to read 
if(0xd800 <= c && c <= 0xdbff && i < len-1) {
Now, if the character is the last one, it gets encoded into utf-8 as is.

I haven't looked at the changes to TextStream yet, Gotta leave for a trip early tomorrow. :)

**Comment by HurryStarfish** [comment]
BlitzMax doesn't support characters over a certain range. There's not much you can do about it.
A smiley has a value of 128512. That simply doesn't fit into 65535.

This surprises me. I always assumed (without ever actually testing it...) that BlitzMax uses UTF-16 encoding for its `String`s. If this isn't the case, it should definitely be switched to use UTF-16, rather than using something obsolete like UCS-2.
Doing his wouldn't change anything about the fact that the code points of a `String` are encoded in `Short`s. UTF-16 requires that code points are stored in *at least* two bytes, which equals one `Short`. Any code point that fits in the 0-65k range (roughly) takes up one `Short`s; any code point larger than that takes up two `Short`s instead.

> So, if you are going to go to that amount of effort, then maybe it's worth more work on top of that to do it properly - for example, should String length return the number of "characters" or the number of "shorts" ?

It would still return the number of `Short`s, for two reasons. The first reason is backwards compatibility. The second reason is the fact that returning the number of characters would require defining what a "character" is, which is actually very difficult for both [regular text](https://en.wikipedia.org/wiki/Combining_character) and [emoji](https://blog.emojipedia.org/emoji-zwj-sequences-three-letters-many-possibilities/). So yeah, you may get a length larger than 1 for things that look like they're only a single "character", [but other languages have the same problem](https://docs.microsoft.com/en-us/dotnet/api/system.char?view=netframework-4.8#char-objects-unicode-characters-and-strings). Stuff like this is probably also the reason why neither Java nor .NET have a `String.Reverse` method...
